libhttpserver/server.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libhttpserver
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"io"
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/hashicorp/golang-lru"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/kbfs/libfs"
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/keybase/kbfs/libmime"
+	"github.com/keybase/kbfs/tlf"
+)
+
+const tokenCacheSize = 64
+const fsCacheSize = 64
+
+// Server is a local HTTP server for serving KBFS content to the front end.
+type Server struct {
+	config libkbfs.Config
+	server *libkb.HTTPSrv
+
+	tokens *lru.Cache
+	fs     *lru.Cache
+}
+
+var _ libkbfs.LocalHTTPServer = (*Server)(nil)
+
+const tokenByteSize = 16
+
+// NewToken implements the libkbfs.LocalHTTPServer interface..
+func (s *Server) NewToken() (token string, err error) {
+	buf := make([]byte, tokenByteSize)
+	if _, err = rand.Read(buf); err != nil {
+		return "", err
+	}
+	token = hex.EncodeToString(buf)
+	s.tokens.Add(token, nil)
+	return token, nil
+}
+
+func (s *Server) handleInvalidToken(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusForbidden)
+	io.WriteString(w, `
+    <html>
+        <head>
+            <title>.kbfs_http_token_invalid</title>
+        </head>
+        <body>
+            token invalid
+        </body>
+    </html>
+    `)
+}
+
+func (s *Server) handleBadRequest(w http.ResponseWriter) {
+	w.WriteHeader(http.StatusBadRequest)
+}
+
+func (s *Server) getHTTPFileSystem(
+	req *http.Request) (toStrip string, fs http.FileSystem, err error) {
+	fields := strings.Split(req.URL.Path, "/")
+	if len(fields) < 3 {
+		return "", nil, errors.New("bad path")
+	}
+	var tlfType tlf.Type
+	switch strings.ToLower(fields[0]) {
+	case "private":
+		tlfType = tlf.Private
+	case "public":
+		tlfType = tlf.Public
+	case "team":
+		tlfType = tlf.SingleTeam
+	default:
+		return "", nil, errors.New("unknown tlf type")
+	}
+
+	tlfHandle, err := libkbfs.GetHandleFromFolderNameAndType(req.Context(),
+		s.config.KBPKI(), s.config.MDOps(), fields[1], tlfType)
+	if err != nil {
+		return "", nil, err
+	}
+
+	toStrip = path.Join(fields[0], fields[1])
+
+	fav := tlfHandle.ToFavorite()
+	if fsCached, ok := s.fs.Get(fav); ok {
+		if tlfFS, ok := fsCached.(*libfs.FS); ok {
+			return toStrip, tlfFS.ToHTTPFileSystem(req.Context()), nil
```

Comment:
Can we cache these `ToHTTPFileSystem` calls?

Reply:
This is merely a wrapper of the underlying `*libfs.FS`. I don't think we can cache its result since it attaches the request's context to it.